Deprecate meshExpansion and add ingress Service port 15012 by default

tbarrella opened the pull request:
As part of #25933. Note that the tcp-dns-tls port is unused

Removing the `ExposeIstiod` test flag is deferred to a future PR to reduce scope for this change

[ ] Configuration Infrastructure
[ ] Docs
[X] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

tbarrella committed:
Deprecate meshExpansion and add ingress Service ports 15012 and 15017 by default

As part of #25933. Note that the tcp-dns-tls port is unused

howardjohn commented on operator/pkg/apis/istio/v1alpha1/validation/validation.go:
```diff
@@ -107,6 +107,8 @@ func deprecatedSettingsMessage(iop *v1alpha1.IstioOperatorSpec) string {
 		{"Values.global.tracer.stackdriver.maxNumberOfMessageEvents", "meshConfig.defaultConfig.tracing.stackdriver.maxNumberOfMessageEvents", 0},
 		{"Values.global.tracer.datadog.address", "meshConfig.defaultConfig.tracing.datadog.address", ""},
+		{"Values.global.meshExpansion.enabled", "Gateway and other Istio networking resources, such as in samples/istiod-gateway/", false},
+		{"Values.gateways.istio-ingressgateway.meshExpansionPorts", "components.ingressGateways[name=istio-ingressgateway].k8s.service.ports", nil},
```
nit: I don't think a warning will help the user much, because I suspect they are not trying to customize the ports but add the normal ones (15012/15017). May be better to just point to the samples as well?

tbarrella replied:
I thought that if they were only using the normal ones, they wouldn't be overriding the value so it would just use the default, so this message wouldn't appear

linsun replied:
do we have the corresponding doc PR to remove the usage of these two configs?

tbarrella replied:
https://github.com/istio/istio.io/pull/7869

howardjohn commented:
Put a hold to give others a chance to review, as I don't think this should be merged after 1hr

cc @linsun @sdake @irisdingbj

nmittler commented on manifests/profiles/default.yaml:
```diff
@@ -64,6 +64,12 @@ spec:
             - port: 443
               targetPort: 8443
               name: https
+            - port: 15012
```
@howardjohn @costinm do we have any concerns about adding these ports by default?

howardjohn replied:
My only concern is if it breaks some LB. GCP ILB used to have 5 port limit which this exceeds, its now removed though.

@nrjpoddar @linsun any idea of other constraints?

howardjohn replied:
Note this *should* be temporary as hopefully we are all on port 443 in the near(ish) future

nmittler replied:
This certainly simplifies the process for exposing istiod (no kubectl patch or bouncing ingress pods required). My concern was also WRT port limits and such. Do you have a rough timeline for the 443 solution?

tbarrella replied:
I could also remove 15017 and add that only in the central istio integration test through `ControlPlaneValues`

linsun replied:
Ok with this, thanks for asking @howardjohn.  I 'm basically running with 15012 and 15017 open before we can figure out switching to use 443.

linsun replied:
Don't you also need to add 15017?

tbarrella replied:
re https://github.com/istio/istio/pull/26146#issuecomment-669775514, I wanted to just add 15012 first to be safer. 15017 could be added easily later if needed

nmittler commented on manifests/profiles/default.yaml:
```diff
@@ -288,12 +294,6 @@ spec:
         zvpn: {}
         env: {}
         meshExpansionPorts:
```
Can we remove "meshExpansionPorts:" too?

tbarrella replied:
Oh yep, and this made me realize they also need to be removed from the subchart...

tbarrella replied:
Nevermind, will keep the subchart values because it's less likely to significantly break things that way. Removing this still

nmittler commented on manifests/profiles/default.yaml:
```diff
@@ -288,12 +294,6 @@ spec:
-        - port: 15012
-          targetPort: 15012
-          name: tcp-istiod
-        - port: 853
```
Do we need 853 exposed as well?

howardjohn replied:
No, it was for something that is not used anymore AFAIK? @costinm can confirm

tbarrella replied:
I searched to see if anything was even listening on that port and couldn't find anything

tbarrella committed:
Remove meshExpansionPorts key from default profile

istio-testing commented:
@tbarrella: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-ipv6-k8s-tests_istio | 398520a9511ce4305a5425a8572367f35f1a3575 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/26146/integ-ipv6-k8s-tests_istio/3826) | `/test integ-ipv6-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@tbarrella: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-ipv6-k8s-tests_istio | 398520a9511ce4305a5425a8572367f35f1a3575 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/26146/integ-ipv6-k8s-tests_istio/3826) | `/test integ-ipv6-k8s-tests_istio`
unit-tests_istio | 398520a9511ce4305a5425a8572367f35f1a3575 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/26146/unit-tests_istio/18251) | `/test unit-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

tbarrella commented:
/retest

irisdingbj commented:
with this change, we will open 15012 and 15017 by default. I recalled we do not want to expose them directly due to some security reason.

Thinking can we only expose them for integration test? if we want to make them default exposed, we can add some logic in operator.go to hack this.

tbarrella commented:
Good point @irisdingbj. I actually prefer to just start with 15012. It's much easier to add 15017 later if needed because it was never present in `meshExpansionPorts`, and relatively harmless, whereas removing it would be a breaking change. Will update (unless someone disagrees)

tbarrella committed:
Remove ingress port 15017 (keeping 15012)

Port 15017 is not necessarily ready to be exposed directly

Also, it's much easier to add 15017 later if needed because it was never
present in meshExpansionPorts, and relatively harmless, whereas removing
it would be a breaking change

tbarrella commented:
/retest

linsun commented:
LGTM, may consider adding 15017 as well.  Would be good to make sure multicluster docs (and/or vm expansion docs) are updated accordingly.

cc @GregHanson @irisdingbj
